client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx

```diff
@@ -953,4 +948,64 @@ function getBusinessPlanByTerm( term: string ) {
 	}
 }
 
+function getRequiredPlan( selectedDesign: Design | undefined, currentPlanSlug: string ) {
+	if ( ! selectedDesign || ! selectedDesign?.design_tier ) {
+		return;
+	}
+	// Different designs require different plans to unlock them, additionally the terms required can vary.
+	// A site with a plan of a given length cannot upgrade a plan of a shorter length. For example,
+	// if a site is on a 2 year starter plan and want to buy an explorer theme, they must buy a 2 year explorer plan
+	// not a 1 year explorer plan.
+	const tierMinimumUpsellPlan =
+		THEME_TIERS[ selectedDesign.design_tier as keyof typeof THEME_TIERS ]?.minimumUpsellPlan;
+
+	let requiredTerm;
+	if ( ! currentPlanSlug || isFreePlan( currentPlanSlug ) ) {
+		// Marketplace themes require upgrading to a monthly business plan or higher, everything else requires an annual plan.
+		if ( selectedDesign?.is_externally_managed ) {
+			requiredTerm = TERM_MONTHLY;
+		} else {
+			requiredTerm = TERM_ANNUALLY;
+		}
+	} else {
+		requiredTerm = getPlan( currentPlanSlug )?.term || '';
+	}
+
+	return getPlanByTerm( tierMinimumUpsellPlan, requiredTerm );
+}
+
+function getPlanByTerm( plan: string, term: string ) {
+	// This is icky, replace later with less hard-coding.
+	switch ( plan ) {
+		case PLAN_PERSONAL:
+			switch ( term ) {
+				case TERM_TRIENNIALLY:
+					return PLAN_PERSONAL_3_YEARS;
+				case TERM_BIENNIALLY:
+					return PLAN_PERSONAL_2_YEARS;
+				case TERM_MONTHLY:
+					return PLAN_PERSONAL_MONTHLY;
+				case TERM_ANNUALLY:
+				default:
+					return PLAN_PERSONAL;
+			}
+		case PLAN_PREMIUM:
+			switch ( term ) {
+				case TERM_TRIENNIALLY:
+					return PLAN_PREMIUM_3_YEARS;
+				case TERM_BIENNIALLY:
+					return PLAN_PREMIUM_2_YEARS;
+				case TERM_MONTHLY:
+					return PLAN_PREMIUM_MONTHLY;
+				case TERM_ANNUALLY:
+				default:
+					return PLAN_PREMIUM;
+			}
+		case PLAN_BUSINESS:
+			return getBusinessPlanByTerm( term );
+		default:
+			return plan;
+	}
```

Comment:
How do you feel about having this as a data structure and querying it instead? I have something like this in mind:
``` 
const youNameIt = {
    PLAN_PERSONAL: {
        TERM_TRIENNIALLY: PLAN_PERSONAL_3_YEARS,
        TERM_BIENNIALLY: PLAN_PERSONAL_2_YEARS,
    .
    .
    .
    PLAN_PREMIUM: {
        TERM_TRIENNIALLY: PLAN_PREMIUM_3_YEARS,
   .
   .
   .
}

function getPlanByTerm( plan: string, term: string ) => youNameIt[plan][term]

```

Reply:
Used `findSimilarPlansKeys` instead which removes the need for either of these approaches and means this component at least, will work if we introduce a four year plan or whatever. Thanks for the nudge to reconsider 👍